src/main/java/io/confluent/connect/jdbc/dialect/CrateDatabaseDialect.java

```diff
@@ -0,0 +1,211 @@
+   *
+   * @param config the connector configuration; may not be null
+   */
+  public CrateDatabaseDialect(AbstractConfig config) {
+    super(config, new IdentifierRules(".", "\"", "\""));
+  }
+
+  /**
+   * Perform any operations on a {@link PreparedStatement} before it is used. This is called from
+   * the {@link #createPreparedStatement(Connection, String)} method after the statement is
+   * created but before it is returned/used.
+   *
+   * <p>This method sets the {@link PreparedStatement#setFetchDirection(int) fetch direction}
+   * to {@link ResultSet#FETCH_FORWARD forward} as an optimization for the driver to allow it to
+   * scroll more efficiently through the result set and prevent out of memory errors.
+   *
+   * @param stmt the prepared statement; never null
+   * @throws SQLException the error that might result from initialization
+   */
+  @Override
+  protected void initializePreparedStatement(PreparedStatement stmt) throws SQLException {
+    log.trace("Initializing PreparedStatement fetch direction to FETCH_FORWARD for '{}'", stmt);
+    stmt.setFetchDirection(ResultSet.FETCH_FORWARD);
+  }
+
+  @Override
+  protected String getSqlType(SinkRecordField field) {
+    if (field.schemaName() != null) {
+      switch (field.schemaName()) {
+        case Decimal.LOGICAL_NAME:
+          return "DOUBLE";
+        case Date.LOGICAL_NAME:
+          return "TIMESTAMP";
+        case Time.LOGICAL_NAME:
+          return "TIMESTAMP";
+        case Timestamp.LOGICAL_NAME:
+          return "TIMESTAMP";
+        default:
+          // fall through to normal types
+      }
+    }
+    switch (field.schemaType()) {
+      case INT8:
+        return "INTEGER";
+      case INT16:
+        return "INTEGER";
+      case INT32:
+        return "INTEGER";
+      case INT64:
+        return "INTEGER";
+      case FLOAT32:
+        return "FLOAT";
+      case FLOAT64:
+        return "FLOAT";
+      case BOOLEAN:
+        return "BOOLEAN";
+      case STRING:
+        return "VARCHAR";
+      case BYTES:
+        return "BYTE";
+      default:
+        return super.getSqlType(field);
+    }
+  }
+
+
+  @Override
+  public String buildUpsertQueryStatement(
+      TableId table,
+      Collection<ColumnId> keyColumns,
+      Collection<ColumnId> nonKeyColumns
+  ) {
+    final Transform<ColumnId> transform = (builder, col) -> {
+      builder.appendIdentifierQuoted(col.name())
+             .append("=EXCLUDED.")
+             .appendIdentifierQuoted(col.name());
+    };
+
+    ExpressionBuilder builder = expressionBuilder();
+    builder.append("INSERT INTO ");
+    builder.append(table);
+    builder.append(" (");
+    builder.appendList()
+           .delimitedBy(",")
+           .transformedBy(ExpressionBuilder.columnNames())
+           .of(keyColumns, nonKeyColumns);
+    builder.append(") VALUES (");
+    builder.appendMultiple(",", "?", keyColumns.size() + nonKeyColumns.size());
+    builder.append(") ON CONFLICT (");
```

Comment:
This doesn't seem to follow the [CrateDB documentation on upserts](https://crate.io/docs/crate/reference/en/latest/general/dml.html?highlight=upsert#upserts-on-duplicate-key-update).

Reply:
I use docs there https://crate.io/docs/crate/reference/en/latest/sql/statements/insert.html#on-conflict-do-update-set. 

`ON CONFLICT DO UPDATE SET` works just like `ON DUPLICATE KEY UPDATE`, but its syntax is slightly different. Additionally, it requires to specify the key columns (conflict target) to perform the duplicate key check against.